atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java

```diff
@@ -514,6 +520,175 @@ public void setOnceTest() {
         keyValueService.truncateTable(userTable);
     }
 
+    @Test
+    public void testMultiCheckAndSetFromEmpty() {
+        Map<Cell, byte[]> updates = ImmutableMap.of(TEST_CELL, val(0, 0));
+
+        MultiCheckAndSetRequest request = MultiCheckAndSetRequest.newCells(TEST_TABLE, TEST_CELL.getRowName(), updates);
+        keyValueService.multiCheckAndSet(request);
+
+        verifyMultiCheckAndSet(updates);
+    }
+
+    @Test
+    public void testMultiCheckAndSetFromOtherValue() {
+        Map<Cell, byte[]> expected = ImmutableMap.of(TEST_CELL, val(0, 0));
+        keyValueService.put(TEST_TABLE, expected, AtlasDbConstants.TRANSACTION_TS);
+
+        Map<Cell, byte[]> updates = ImmutableMap.of(TEST_CELL, val(0, 1));
+        MultiCheckAndSetRequest request =
+                MultiCheckAndSetRequest.multipleCells(TEST_TABLE, TEST_CELL.getRowName(), expected, updates);
+        keyValueService.multiCheckAndSet(request);
+
+        verifyMultiCheckAndSet(updates);
+    }
+
+    @Test
+    public void testMultiCheckAndSetAndBackAgain() {
+        testMultiCheckAndSetFromOtherValue();
+
+        Map<Cell, byte[]> expected = ImmutableMap.of(TEST_CELL, val(0, 1));
+        Map<Cell, byte[]> updates = ImmutableMap.of(TEST_CELL, val(0, 0));
+
+        MultiCheckAndSetRequest request =
+                MultiCheckAndSetRequest.multipleCells(TEST_TABLE, TEST_CELL.getRowName(), expected, updates);
+        keyValueService.multiCheckAndSet(request);
+
+        verifyMultiCheckAndSet(updates);
+    }
+
+    @Test
+    public void testMultiCheckAndSetLargeValue() {
+        byte[] megabyteValue = new byte[1048576];
+        Map<Cell, byte[]> updates = ImmutableMap.of(TEST_CELL, megabyteValue);
+
+        MultiCheckAndSetRequest request = MultiCheckAndSetRequest.newCells(TEST_TABLE, TEST_CELL.getRowName(), updates);
+        keyValueService.multiCheckAndSet(request);
+    }
+
+    @Test
+    public void testMultiCheckAndSetFromValueWhenNoValue() {
+        Map<Cell, byte[]> expected = ImmutableMap.of(TEST_CELL, val(0, 0));
+        Map<Cell, byte[]> updates = ImmutableMap.of(TEST_CELL, val(0, 1));
+        MultiCheckAndSetRequest request =
+                MultiCheckAndSetRequest.multipleCells(TEST_TABLE, TEST_CELL.getRowName(), expected, updates);
+        assertThatThrownBy(() -> keyValueService.multiCheckAndSet(request))
```

Comment:
Here and maybe in a couple additional tests: we want to check that the actual values in the exception are correct